### PR TITLE
Updated scalea files using  the latest DUH

### PR DESCRIPTION
**Pull request by sistillson**
added examples of updating device tree

### CODE REVIEW

File: craft/pio/src/pio.scala [hunk 1]
```diff
@@ -36,9 +36,13 @@ class Lpio(c: pioParams)(implicit p: Parameters) extends LpioBase(c)(p)
 {
 
 // User code here
+  override def extraResources(resources: ResourceBindings) =
+      Map("pio-width" -> Seq(ResourceInt(pioWidth)),
+          "compatible" -> Seq(ResourceString("sifive,blenderific111")))
```
Comment (richardxia) [inline craft/pio/src/pio.scala]: You shouldn't set the `compatible` property here, since that's what the second argument to `SimpleDevice()` is supposed to be.

@albertchen-sifive, we might want to add a runtime check to the base classes that prevents setting the `compatible` via this map.

File: craft/loopback/src/loopback.scala [hunk 2]
```diff
@@ -29,12 +29,21 @@ class Lloopback(c: loopbackParams)(implicit p: Parameters) extends LloopbackBase
 
 // User code here
 
+
+
+
+  override def extraResources(resources: ResourceBindings) =
+      Map("color" -> Seq(ResourceString("black"), ResourceString("White")),
+           "bob" -> Seq(ResourceInt(7))
+      )
```
Comment (richardxia) [inline craft/loopback/src/loopback.scala]: Can you revert these changes, which are presumably tests that weren't meant to be merged in?

**Review by richardxia — COMMENTED** [review]
I have a few comments, but this mostly looks good to me.

File: craft/pio/src/pio.scala (hunk 1 above)
Comment (sistillson) [reply craft/pio/src/pio.scala]: fixed

File: craft/loopback/src/loopback.scala (hunk 2 above)
Comment (sistillson) [reply craft/loopback/src/loopback.scala]: gone

File: craft/pio/src/pio.scala (hunk 1 above)
Comment (albertchen-sifive) [reply craft/pio/src/pio.scala]: > @albertchen-sifive, we might want to add a runtime check to the base classes that prevents setting the compatible via this map.

I think we do want people to be able to override this though? does adding the `"compatible"` mapping override any existing mapping or append to it?

Comment (sistillson) [reply craft/pio/src/pio.scala]: It overwrites.

Comment (richardxia) [reply craft/pio/src/pio.scala]: You make a good point about allowing people to override it. We'll just need to make sure our documentation in this is clear about the default value that it gets.

File: craft/pio/src/pio.scala [hunk 3]
```diff
@@ -36,9 +36,13 @@ class Lpio(c: pioParams)(implicit p: Parameters) extends LpioBase(c)(p)
 {
 
 // User code here
+  override def extraResources(resources: ResourceBindings) =
+      Map("pio-width" -> Seq(ResourceInt(pioWidth)),
+          "pio-type" -> Seq(ResourceString("loopback")))
```
Comment (richardxia) [inline craft/pio/src/pio.scala]: This "pio-type" isn't really a property of the PIO itself. That's describing the fact that we've put a loopback on it in our test harness.

**Review by richardxia — APPROVED** [review]
LGTM, aside from one last comment about that `pio-type` field you added.

File: craft/pio/src/pio.scala (hunk 3 above)
Comment (sistillson) [reply craft/pio/src/pio.scala]: I was trying to come up with a fake example to show how it could be used. If that is ok, I will merge

File: craft/loopback/src/loopback.scala [hunk 4]
```diff
@@ -31,10 +31,11 @@ class Lloopback(c: loopbackParams)(implicit p: Parameters) extends LloopbackBase
 
 }
 
-class NloopbackTop(c: NloopbackTopParams)(implicit p: Parameters) extends NloopbackTopBase(c)(p)
+class NloopbackTop(c: NloopbackTopParams)(implicit p: Parameters) extends NloopbackTopBase(c)(p) with BindingScope
 {
 
 // User code here
+  ResourceBinding { Resource(imp.device, "exists").bind(ResourceString("yes")) }
```
Comment (albertchen-sifive) [inline craft/loopback/src/loopback.scala]: @richardxia @sistillson I think this is something we probably want to always emit in the base code?

**Review by albertchen-sifive — APPROVED** [review]

File: craft/loopback/src/loopback.scala (hunk 4 above)
Comment (richardxia) [reply craft/loopback/src/loopback.scala]: What exactly does this do? I see this [exact line](https://github.com/chipsalliance/rocket-chip/blob/1f561026f85b885a7b13c3a5dffda6a47fa2d6bb/src/main/scala/diplomacy/Resources.scala#L89) in `DeviceSnippet` in rocket-chip.

`DeviceSnippet` itself, which is an abstract class, seems to only be used to define and immediately instantiate an  anonymous class which overrides the `describe()` method. It feels like we should at the very least move this to the base code emitted by duh-scala, but we may also want to use `DeviceSnippet` itself.

File: craft/pio/src/pio.scala (hunk 3 above)
Comment (richardxia) [reply craft/pio/src/pio.scala]: I'm afraid that this example may cause more confusion than it provides help, since it may cause people to think that this is actually linked somehow to the other hardware block in this repo with the name "loopback".

I'd rather leave out fake examples, and instead come up with real examples. This is a perfectly usable, real IP block, albeit simple, and there's nothing preventing someone from actually using this block in a real design, so I don't think we should pollute it with fake examples. If we want to provide other examples of resource types, I think we can include it in documentation we have on this (and how the Resources system works in general), but I don't think we necessarily need to include them all here. There are examples of other Resource types throughout rocket-chip.

I would strongly request that we remove this outright, or that we at least replace it with some other property that actually describes this block.
